src/Bearded.TD/Game/Simulation/Elements/ArcTree.cs

```diff
@@ -0,0 +1,171 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Bearded.TD.Game.Simulation.GameObjects;
+using Bearded.TD.Game.Simulation.Navigation;
+using Bearded.TD.Game.Simulation.Weapons;
+using Bearded.TD.Tiles;
+using Bearded.TD.Utilities.Collections;
+using Bearded.Utilities.Linq;
+using Bearded.Utilities.SpaceTime;
+
+namespace Bearded.TD.Game.Simulation.Elements;
+
+interface IArcSource
+{
+    IEnumerable<IArcModifier> ContinuationModifiers { get; }
+
+    ArcTree.Continuation GetInitialArcParametersWithoutModifiers();
+}
+
+interface IArcModifier
+{
+    void Modify(ref ArcTree.Continuation arc);
+}
+
+static class ArcTree
+{
+    public record struct Arc(
+        GameObject Source,
+        GameObject Target,
+        Continuation Continuation
+    );
+
+    public record struct Continuation(
+        GameObject Source,
+        int BouncesBefore,
+        int BouncesLeft,
+        int Branches,
+        int MaxBounceDistance)
+    {
+        public static Continuation Initial(
+            GameObject source, int bounces = 1, int branches = 1, int maxBounceDistance = 2)
+            => new(source, 0, bounces, branches, maxBounceDistance);
+    }
+
+    public static List<Arc> Strike(
+        GameState game,
+        GameObject? source,
+        IArcSource arcSource,
+        ImmutableArray<Tile> areaInRange)
+    {
+        var state = setupState(game, source, arcSource);
+
+        var initialContinuation = createInitialContinuation(state, arcSource);
+        connectToTargetsInArea(state, initialContinuation, areaInRange);
+
+        while (state.QueuedContinuations.TryDequeue(out var continuation))
+        {
+            var area = getAreaForContinuation(state, continuation);
+            connectToTargetsInArea(state, continuation, area);
+        }
```

Comment:
Compared to before, I think the similarity between these two sections shows more clearly why it can't fully reuse the same code, while clearly highlighting the part it does reuse on a high level instead of hidden somewhere deep in the algorithm.

Reply:
The only thing that I think isn't quite obvious is that `connectToTargetsInArea` adds further continuations. Or rather: it obviously must be doing so, since that's the recurring thing, but it's maybe not clear why it would exactly. Connecting implies continuing here, and I'm not sure how to represent that better since it happens at the same time in the low-level algorithm below. Calling the method `connectToTargetsInAreaAndQueueContinuations` is possible, but... ugly. Perhaps there's a further refactor that can make that unnecessary or clearer in some other way, but I'm leaving it as is for now.